common/src/main/java/org/schemarepo/AbstractBackendRepository.java

```diff
@@ -0,0 +1,104 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package org.schemarepo;
+
+/**
+ * Parent class of the actual backend (as opposed to decorating) repositories.
+ * Each backend repository is expected to cache and validate its subjects.
+ */
+public abstract class AbstractBackendRepository extends BaseRepository {
+
+  protected final InMemorySubjectCache subjectCache = new InMemorySubjectCache();
+  protected final ValidatorFactory validators;
+
+  protected AbstractBackendRepository(final ValidatorFactory validators) {
+    this.validators = validators != null ? validators : ValidatorFactory.EMPTY;
+  }
+
+  /**
+   * Concrete subclasses must override to return specific implementation of {@link org.schemarepo.Subject}.
+   * Note that this instantiates the "right" subject only, any "side effects" (such as mutating persistent state)
+   * must be implemented in {@link #registerSubjectInBackend(String, SubjectConfig)}
+   * @param subjectName subject name
+   * @param config subject config
+   * @return Subject
+   */
+  protected abstract Subject instantiateSubject(final String subjectName, final SubjectConfig config);
+
+  /**
+   * Creates, applies validation decorator, and caches subject.
+   * @param subjectName subject name
+   * @param config subject config
+   * @return Subject the newly created instance or possibly pre-existing cached instance
+   */
+  protected final Subject createAndCacheSubject(final String subjectName, final SubjectConfig config) {
+    return subjectCache.add(Subject.validatingSubject(instantiateSubject(subjectName, config), validators));
+  }
+
+  @Override
+  public synchronized Subject register(final String subjectName, final SubjectConfig config) {
+    isValid();
+    Subject subject = subjectCache.lookup(subjectName);
+    if (subject == null) {
+      registerSubjectInBackend(subjectName, config);
+      subject = createAndCacheSubject(subjectName, config);
+    } else {
+      logger.debug("Subject {} already exists, reusing", subjectName);
+    }
+    return subject;
+  }
+
+  /**
+   * Backend-specific registration logic. This is called during registration upon cache miss.
+   * Usually is overridden by subclasses. By default, do nothing.
+   * @param subjectName subject name
+   * @param config subject config
+   */
+  protected void registerSubjectInBackend(final String subjectName, final SubjectConfig config) {
+  }
+
+  @Override
+  public synchronized Subject lookup(final String subjectName) {
+    isValid();
+    Subject subject = subjectCache.lookup(subjectName);
+    if (subject == null) {
+      if (checkSubjectExistsInBackend(subjectName)) {
+        subject = createAndCacheSubject(subjectName, null);
```

Comment:
I am tempted to say that  changing the signature to ```instantiateSubject(String subjectName)``` would be ideal.

Forcing this signature would eliminate the inconsistencies and more clearly define the responsibilities of each function.

This would require changing the ```InMemoryRepository``` so that ```registerSubjectInBackend``` actually does something. Then, the ```MemSubject``` constructor could interrogate the ```subjectCache``` of the ```InMemoryRepository``` it is contained within, and bootstrap its state from there.